pkg/reconciler/v1alpha1/pipelinerun/resources/pipelinestate.go

```diff
@@ -0,0 +1,163 @@
+package resources
+
+import (
+	"fmt"
+
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+// GetNextTask returns the next Task for which a TaskRun should be created,
+// or nil if no TaskRun should be created.
+func GetNextTask(prName string, state []*PipelineRunTaskRun, logger *zap.SugaredLogger) *PipelineRunTaskRun {
+	for _, prtr := range state {
+		if prtr.TaskRun != nil {
+			c := prtr.TaskRun.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
+			if c == nil {
+				logger.Infof("TaskRun %s doesn't have a condition so it is just starting and we shouldn't start more for PipelineRun %s", prtr.TaskRunName, prName)
+				return nil
+			}
+			switch c.Status {
+			case corev1.ConditionFalse:
+				logger.Infof("TaskRun %s has failed; we don't need to run PipelineRun %s", prtr.TaskRunName, prName)
+				return nil
+			case corev1.ConditionUnknown:
+				logger.Infof("TaskRun %s is still running so we shouldn't start more for PipelineRun %s", prtr.TaskRunName, prName)
+				return nil
+			}
+		} else if canTaskRun(prtr.PipelineTask) {
+			logger.Infof("TaskRun %s should be started for PipelineRun %s", prtr.TaskRunName, prName)
+			return prtr
+		}
+	}
+	logger.Infof("No TaskRuns to start for PipelineRun %s", prName)
+	return nil
+}
+
+func canTaskRun(pt *v1alpha1.PipelineTask) bool {
+	// Check if Task can run now. Go through all the input constraints
+	return true
+}
+
+// PipelineRunTaskRun contains a Task and its associated TaskRun, if it
+// exists. TaskRun can be nil to represent there being no TaskRun.
+type PipelineRunTaskRun struct {
+	Task         *v1alpha1.Task
+	PipelineTask *v1alpha1.PipelineTask
+	TaskRunName  string
+	TaskRun      *v1alpha1.TaskRun
+}
+
+// GetTask is a function that will retrieve the Task name from namespace.
+type GetTask func(namespace, name string) (*v1alpha1.Task, error)
+
+// GetTaskRun is a function that will retrieve the TaskRun name from namespace.
+type GetTaskRun func(namespace, name string) (*v1alpha1.TaskRun, error)
+
+// GetPipelineState retrieves all Tasks instances which the pipeline p references, getting
+// instances from getTask. It will also check if there is a corresponding TaskRun for the
+// Task using getTaskRun (the name is built from pipelineRunName). If it is unable to
+// retrieve an instance of a referenced Task, it will return an error, otherwise it
+// returns a list of all of the Tasks retrieved, and their TaskRuns if applicable.
+func GetPipelineState(getTask GetTask, getTaskRun GetTaskRun, p *v1alpha1.Pipeline, pipelineRunName string) ([]*PipelineRunTaskRun, error) {
+	state := []*PipelineRunTaskRun{}
+	for i := range p.Spec.Tasks {
+		pt := p.Spec.Tasks[i]
+		t, err := getTask(p.Namespace, pt.TaskRef.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get tasks for Pipeline %q: Error getting task %q : %s",
```

Comment:
Now, this is case where we need to check if Task does not exists.
If it does not exist, then we should quit re-conciling this pipeline since its invalid and nothing can be done.  
For that to happen, we should return "nil" from the `pipelinerun.Reconciler.reconcile`

however, there was some other error while fetching task then we keep trying. 
Does that make sense? 

I would say, we can create a InvalidPipelineError and return that from here when a Task is not found.
In the `reconcile` method, we should if err is something other than InvalidPipelineError and return nil if its InvalidPipelineError

Reply:
yep that makes sense! okay ill update this so that if we can't find a task, we return nil from the reconcile loop and stop reconciling.

thanks for catching this and the detailed explanation!